File: src/main/java/com/example/seatrans/features/post/service/PostService.java
package com.example.seatrans.features.post.service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.example.seatrans.features.auth.model.User;
import com.example.seatrans.features.auth.repository.UserRepository;
import com.example.seatrans.features.post.dto.CategoryResponse;
import com.example.seatrans.features.post.dto.PostRequest;
import com.example.seatrans.features.post.dto.PostResponse;
import com.example.seatrans.features.post.model.Category;
import com.example.seatrans.features.post.model.Post;
import com.example.seatrans.features.post.model.PostCategory;
import com.example.seatrans.features.post.model.PostImage;
import com.example.seatrans.features.post.repository.CategoryRepository;
import com.example.seatrans.features.post.repository.PostImageRepository;
import com.example.seatrans.features.post.repository.PostRepository;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

/**
 * Service for Post management
 */
@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class PostService {
    
    private final PostRepository postRepository;
    private final UserRepository userRepository;
    private final PostImageRepository postImageRepository;
    private final CategoryRepository categoryRepository;
    
    /**
     * Create a new post
     */
    public PostResponse createPost(PostRequest request, String email) {
        log.info("Creating post with title: {}", request.getTitle());
        
        User author = userRepository.findByEmail(email)
            .orElseThrow(() -> new RuntimeException("Author not found"));
        
        Post post = Post.builder()
            .title(request.getTitle())
            .content(request.getContent())
            .author(author)
            .thumbnailUrl(request.getThumbnailUrl())
            .isPublished(request.getIsPublished() != null ? request.getIsPublished() : false)
            .build();
        
        if (post.getIsPublished()) {
            post.setPublishedAt(LocalDateTime.now());
        }
        
        Post savedPost = postRepository.save(post);
        
        // Handle categories
        if (request.getCategoryIds() != null && !request.getCategoryIds().isEmpty()) {
            for (Long categoryId : request.getCategoryIds()) {
                Category category = categoryRepository.findById(categoryId)
                    .orElseThrow(() -> new RuntimeException("Category not found: " + categoryId));
                PostCategory postCategory = PostCategory.builder()
                    .post(savedPost)
                    .category(category)
                    .build();
                savedPost.getPostCategories().add(postCategory);
            }
            savedPost = postRepository.save(savedPost);
        }
        
        log.info("Post created with ID: {}", savedPost.getId());
        
        return toResponse(savedPost);
    }
    
    /**
     * Update an existing post
     */
    public PostResponse updatePost(Long id, PostRequest request) {
        log.info("Updating post with ID: {}", id);
        
        Post post = postRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Post not found with id: " + id));
        
        post.setTitle(request.getTitle());
        post.setContent(request.getContent());
        post.setThumbnailUrl(request.getThumbnailUrl());
        
        // Handle publish status change
        if (request.getIsPublished() != null && request.getIsPublished() != post.getIsPublished()) {
            if (request.getIsPublished()) {
                post.publish();
            } else {
                post.unpublish();
            }
        }
        
        // Update categories
        if (request.getCategoryIds() != null) {
            // Clear existing categories
            post.getPostCategories().clear();
            
            // Add new categories
            for (Long categoryId : request.getCategoryIds()) {
                Category category = categoryRepository.findById(categoryId)
                    .orElseThrow(() -> new RuntimeException("Category not found: " + categoryId));
                PostCategory postCategory = PostCategory.builder()
                    .post(post)
                    .category(category)
                    .build();
                post.getPostCategories().add(postCategory);
            }
        }
        
        Post updatedPost = postRepository.save(post);
        log.info("Post updated: {}", updatedPost.getId());
        
        return toResponse(updatedPost);
    }
    
    /**
     * Delete a post
     */
    public void deletePost(Long id) {
        log.info("Deleting post with ID: {}", id);
        
        if (!postRepository.existsById(id)) {
            throw new RuntimeException("Post not found with id: " + id);
        }
        
        postRepository.deleteById(id);
        log.info("Post deleted: {}", id);
    }
    
    /**
     * Get post by ID
     */
    public PostResponse getPostById(Long id) {
        Post post = postRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Post not found with id: " + id));
        
        return toResponse(post);
    }
    
    /**
     * Get post by ID and increment view count
     */
    public PostResponse getPostByIdWithViewCount(Long id) {
        Post post = postRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Post not found with id: " + id));
        
        post.incrementViewCount();
        postRepository.save(post);
        
        return toResponse(post);
    }
    
    /**
     * Get all posts (for admin)
     */
    public List<PostResponse> getAllPosts() {
        List<Post> posts = postRepository.findAll(Sort.by(Sort.Direction.DESC, "createdAt"));
        return posts.stream()
            .map(this::toResponse)
            .collect(Collectors.toList());
    }
    
    /**
     * Get all published posts
     */
    public List<PostResponse> getAllPublishedPosts() {
        List<Post> posts = postRepository.findByIsPublishedTrueOrderByPublishedAtDesc();
        return posts.stream()
            .map(this::toResponse)
            .collect(Collectors.toList());
    }
    
    /**
     * Get published posts with pagination
     */
    public Page<PostResponse> getPublishedPostsPaginated(int page, int size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "publishedAt"));
        Page<Post> posts = postRepository.findByIsPublishedTrue(pageable);
        return posts.map(this::toResponse);
    }
    
    /**
     * Get posts by category
     */
    public List<PostResponse> getPostsByCategory(String category) {
        List<Post> posts = postRepository.findByIsPublishedTrueAndCategoryOrderByPublishedAtDesc(category);
        return posts.stream()
            .map(this::toResponse)
            .collect(Collectors.toList());
    }
    
    /**
     * Search posts by keyword
     */
    public List<PostResponse> searchPosts(String keyword) {
        List<Post> posts = postRepository.searchPublishedByKeyword(keyword);
        return posts.stream()
            .map(this::toResponse)
            .collect(Collectors.toList());
    }
    
    /**
     * Get latest published posts
     */
    public List<PostResponse> getLatestPosts(int limit) {
        Pageable pageable = PageRequest.of(0, limit);
        List<Post> posts = postRepository.findLatestPublished(pageable);
        return posts.stream()
            .map(this::toResponse)
            .collect(Collectors.toList());
    }
    
    /**
     * Publish a post
     */
    public PostResponse publishPost(Long id) {
        Post post = postRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Post not found with id: " + id));
        
        post.publish();
        Post publishedPost = postRepository.save(post);
        log.info("Post published: {}", publishedPost.getId());
        
        return toResponse(publishedPost);
    }
    
    /**
     * Unpublish a post
     */
    public PostResponse unpublishPost(Long id) {
        Post post = postRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Post not found with id: " + id));
        
        post.unpublish();
        Post unpublishedPost = postRepository.save(post);
        log.info("Post unpublished: {}", unpublishedPost.getId());
        
        return toResponse(unpublishedPost);
    }

    /**
     * Save post image
     */
    public void savePostImage(Long postId, String url) {
        if (postId != null) {
            Post post = postRepository.findById(postId)
                .orElseThrow(() -> new RuntimeException("Post not found with id: " + postId));
            
            PostImage postImage = PostImage.builder()
                .post(post)
                .url(url)
                .build();
            
            postImageRepository.save(postImage);
        } else {
            // If no post ID, we just save the image record without a post link for now
            // Or we could just skip saving to DB if we only care about linked images
            PostImage postImage = PostImage.builder()
                .url(url)
                .build();
            postImageRepository.save(postImage);
        }
    }
    
    /**
     * Convert Post entity to PostResponse DTO
     */
    private PostResponse toResponse(Post post) {
        List<CategoryResponse> categories = post.getPostCategories().stream()
            .map(pc -> CategoryResponse.builder()
                .id(pc.getCategory().getId())
                .name(pc.getCategory().getName())
                .description(pc.getCategory().getDescription())
                .build())
            .collect(Collectors.toList());
        
        return PostResponse.builder()
            .id(post.getId())
            .title(post.getTitle())
            .content(post.getContent())
            .authorId(post.getAuthor().getId())
            .authorName(post.getAuthor().getFullName() != null ? post.getAuthor().getFullName() : post.getAuthor().getEmail())
            .categories(categories)
            .thumbnailUrl(post.getThumbnailUrl())
            .publishedAt(post.getPublishedAt())
            .isPublished(post.getIsPublished())
            .viewCount(post.getViewCount())
            .createdAt(post.getCreatedAt())
            .updatedAt(post.getUpdatedAt())
            .build();
    }
}

File: frontend/src/modules/posts/services/postService.ts
import { apiClient } from '@/shared/utils/apiClient'
import { API_CONFIG } from '@/shared/config/api.config'
import type { ApiResponse } from '@/shared/types/api.types'

export interface CategoryResponse {
  id: number
  name: string
  description?: string
  createdAt?: string
}

export interface Post {
  id: number
  title: string
  content: string
  summary?: string
  authorId: number
  authorName: string
  author?: { fullName?: string }
  categories: CategoryResponse[]
  tags?: string[]
  readingTime?: number
  thumbnailUrl?: string
  publishedAt?: string
  isPublished: boolean
  viewCount: number
  createdAt: string
  updatedAt: string
}

export interface PostRequest {
  title: string
  content: string
  categoryIds?: number[]
  thumbnailUrl?: string
  isPublished?: boolean
}

const mapPost = (raw: any): Post => {
  const words = typeof raw?.content === 'string' ? raw.content.split(/\s+/).length : 0
  const readingTime = raw?.readingTime ?? Math.max(1, Math.round(words / 200))
  return {
    ...raw,
    categories: Array.isArray(raw?.categories) ? raw.categories : [],
    tags: Array.isArray(raw?.tags) ? raw.tags : [],
    summary: raw?.summary ?? '',
    author: { fullName: raw?.authorFullName || raw?.authorName },
    readingTime,
  }
}

export const postService = {
  // Admin endpoints
  getAllPosts: async (): Promise<Post[]> => {
    const response = await apiClient.get<ApiResponse<Post[]>>(API_CONFIG.POSTS.ADMIN_BASE)
    const result = await response.json()
    return Array.isArray(result.data) ? result.data.map(mapPost) : []
  },

  getPostById: async (id: number): Promise<Post> => {
    const response = await apiClient.get<ApiResponse<Post>>(API_CONFIG.POSTS.ADMIN_BY_ID(id))
    const result = await response.json()
    return mapPost(result.data)
  },

  createPost: async (postData: PostRequest): Promise<Post> => {
    const response = await apiClient.post<ApiResponse<Post>>(API_CONFIG.POSTS.ADMIN_BASE, postData)

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.message || 'Failed to create post')
    }

    const result: ApiResponse<Post> = await response.json()
    return mapPost(result.data)
  },

  updatePost: async (id: number, postData: PostRequest): Promise<Post> => {
    const response = await apiClient.put<ApiResponse<Post>>(API_CONFIG.POSTS.ADMIN_BY_ID(id), postData)

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.message || 'Failed to update post')
    }

    const result: ApiResponse<Post> = await response.json()
    return mapPost(result.data)
  },

  deletePost: async (id: number): Promise<void> => {
    const response = await apiClient.delete(API_CONFIG.POSTS.ADMIN_BY_ID(id))

    if (!response.ok) {
      throw new Error('Failed to delete post')
    }
  },

  publishPost: async (id: number): Promise<Post> => {
    const response = await apiClient.post<ApiResponse<Post>>(API_CONFIG.POSTS.PUBLISH(id))

    if (!response.ok) {
      throw new Error('Failed to publish post')
    }

    const result: ApiResponse<Post> = await response.json()
    return mapPost(result.data)
  },

  unpublishPost: async (id: number): Promise<Post> => {
    const response = await apiClient.post<ApiResponse<Post>>(API_CONFIG.POSTS.UNPUBLISH(id))

    if (!response.ok) {
      throw new Error('Failed to unpublish post')
    }

    const result: ApiResponse<Post> = await response.json()
    return mapPost(result.data)
  },

  getPublishedPosts: async (category?: string, search?: string): Promise<Post[]> => {
    const params = new URLSearchParams()
    if (category) params.append('category', category)
    if (search) params.append('search', search)

    const url = `${API_CONFIG.POSTS.PUBLIC_BASE}${params.toString() ? `?${params.toString()}` : ''}`
    const response = await apiClient.get<ApiResponse<Post[]>>(url, { skipAuth: true })

    if (!response.ok) {
      const errorText = await response.text()
      console.error('Failed to fetch posts:', errorText)
      throw new Error(`Failed to fetch published posts: ${response.status} ${response.statusText}`)
    }

    const result: ApiResponse<Post[]> = await response.json()

    if (!result.success || !result.data) {
      throw new Error(result.message || 'Invalid response from server')
    }

    return Array.isArray(result.data) ? result.data.map(mapPost) : []
  },

  getPublicPostById: async (id: number): Promise<Post> => {
    const response = await apiClient.get<ApiResponse<Post>>(API_CONFIG.POSTS.PUBLIC_BY_ID(id), { skipAuth: true })

    if (!response.ok) {
      throw new Error('Failed to fetch post')
    }

    const result: ApiResponse<Post> = await response.json()
    return mapPost(result.data)
  },

  // alias for clarity with ArticleDetailPage
  getById: async (id: number): Promise<Post> => {
    return postService.getPublicPostById(id)
  },

  getLatestPosts: async (limit: number = 5): Promise<Post[]> => {
    const response = await apiClient.get<ApiResponse<Post[]>>(
      `${API_CONFIG.POSTS.LATEST}?limit=${limit}`,
      { skipAuth: true }
    )

    if (!response.ok) {
      throw new Error('Failed to fetch latest posts')
    }

    const result: ApiResponse<Post[]> = await response.json()
    return Array.isArray(result.data) ? result.data.map(mapPost) : []
  },

  uploadImage: async (file: File, postId?: number): Promise<string> => {
    const formData = new FormData()
    formData.append('file', file)
    if (postId) {
      formData.append('postId', postId.toString())
    }

    const response = await apiClient.post<ApiResponse<string>>(
      API_CONFIG.POSTS.UPLOAD_IMAGE,
      formData
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.message || 'Failed to upload image')
    }

    const result: ApiResponse<string> = await response.json()
    return `${API_CONFIG.ASSET_BASE_URL}${result.data}`
  },
}
